Policy Enforcement Fixes

ianvernon opened the pull request:
This commit addresses various issues with respect to Policy Enforcement, specifically those mentioned in the below tickets. 

Policy enforcement mode is now configurable via the CLI in the following manner:
`cilium config PolicyEnforcement={always, default, enabled}`. 

Before, it was: `cilium config Policy={Enabled, Disabled}`

This meant that it was  impossible to switch Policy Enforcement mode back to the 'default' mode (for more information w.r.t. the specifics for policy enforcement see issue #788 ). 

Extensive tests have been added to ensure that the configuration updates for policy enforcement are reflected in the endpoints managed by Cilium. 

Fixes: #1298 
Fixes: #1291 
Fixes: #1082

aanm commented on cilium/cmd/helpers.go:
```diff
@@ -20,6 +20,7 @@ import (
 
 	"github.com/cilium/cilium/pkg/endpoint"
 
+	"github.com/cilium/cilium/pkg/policy"
```
should be one line above

ianvernon replied:
Fixed.

aanm commented on cilium/cmd/endpoint_list.go:
```diff
@@ -44,7 +44,7 @@ func init() {
 func listEndpoint(w *tabwriter.Writer, ep *models.Endpoint, id string, label string) {
 	var isPolicyEnabled string
-	if ep.PolicyEnabled {
+	if *ep.PolicyEnabled {
```
is this really needed?

tgraf replied:
This is not safe. If it's a pointer then swagger will not guarantee that it is always set.

You can do:
```
import (
             "github.com/go-openapi/swag"
)

if swag.BoolValue(ep.PolicyEnabled) {
```

ianvernon replied:
Fixed.

aanm commented on cilium/cmd/config.go:
```diff
@@ -22,9 +22,10 @@ import (
 	"github.com/cilium/cilium/daemon/options"
-	"github.com/cilium/cilium/pkg/option"
 
+	"github.com/cilium/cilium/pkg/option"
```
imports out of order

ianvernon replied:
Fixed.

aanm commented on cilium/cmd/config.go:
```diff
@@ -77,12 +78,26 @@ func configDaemon(cmd *cobra.Command, opts []string) {
 		dumpConfig(resp.Configuration.Mutable)
 		fmt.Printf("%-24s %s\n", "k8s-configuration", resp.K8sConfiguration)
 		fmt.Printf("%-24s %s\n", "k8s-endpoint", resp.K8sEndpoint)
+		fmt.Printf("%-24s %s\n", "PolicyEnforcement", resp.PolicyEnforcement)
```
We already have a `Policy` flag. What will be the meaning of this one? I think we should rename it to `PolicyEnforcementMode` or `PolicyMode`
```
$ cilium config
Conntrack                Enabled
ConntrackAccounting      Enabled
ConntrackLocal           Disabled
Debug                    Enabled
DropNotification         Enabled
Policy                   Disabled
PolicyTracing            Disabled
k8s-configuration        /var/lib/cilium/cilium.kubeconfig
k8s-endpoint             http://
```

tgraf replied:
Can we just remove the old Policy know given we put a more powerful knob in place?

ianvernon replied:
I'll see if I can remove the old Policy boolean option entirely from the daemon. I only think it is used in the daemon in `daemon/daemon.go:writeNetdevHeader()`. I can create a function that writes the `#define` for PolicyEnforcement for the daemon.

ianvernon replied:
Done.

aanm commented on daemon/daemon.go:
```diff
@@ -244,11 +244,9 @@ func (d *Daemon) AlwaysAllowLocalhost() bool {
-
 func (d *Daemon) PolicyEnforcement() string {
+	d.conf.EnablePolicyMU.RLock()
+	defer d.conf.EnablePolicyMU.RUnlock()
```
`defer d.conf.EnablePolicyMU.RUnlock()` creates overhead for a small function like this, you can write it as 
```
func (d *Daemon) PolicyEnforcement() (pe string) {
     d.conf.EnablePolicyMU.RLock()
     pe = d.conf.EnablePolicy
     d.conf.EnablePolicyMU.RUnlock()
     return
}
```

aanm reviewed (commented):
Note to self: reviewed until `daemon: PolicyEnforcement fixes`

tgraf commented on api/v1/openapi.yaml:
```diff
@@ -805,6 +808,20 @@ definitions:
         type: array
         items:
           type: string
+  L3Policy:
```
Can we rename this to CIDRPolicy? As I understand this, this will only contain the CIDR level policy rules

ianvernon replied:
Done.

tgraf commented on api/v1/openapi.yaml:
```diff
@@ -805,6 +808,20 @@ definitions:
+    type: object
+    properties:
+      ingress:
+        description: list of L3 Ingress rules
```
Description should also make it clear that these are CIDR rules

ianvernon replied:
Done.

tgraf commented on daemon/daemon.go:
```diff
@@ -1086,23 +1084,56 @@ func (h *patchConfig) Handle(params PatchConfigParams) middleware.Responder {
+	if enforcement != "" {
+		switch enforcement {
+		case endpoint.NeverEnforce, endpoint.DefaultEnforcement, endpoint.AlwaysEnforce:
+			// If the policy enforcement configuration has indeed changed, we have
```
You have a race condition here. Because you unlock `EnablePolicyMU` and keep the variable it may change by the time you compare it.

I suggest you lock here...

ianvernon replied:
Fixed - see comment on unlocking of the mutex as well.

tgraf commented on daemon/daemon.go:
```diff
@@ -1086,23 +1084,56 @@ func (h *patchConfig) Handle(params PatchConfigParams) middleware.Responder {
+				_, policyEnforcementChanged = d.EnablePolicyEnforcement()
+				d.GetPolicyRepository().Mutex.RUnlock()
+				d.TriggerPolicyUpdates([]policy.NumericIdentity{})
+			}
```
... and unlock here

ianvernon replied:
I fixed the race condition, but not deferring the unlocking of `EnablePolicyMU` introduced very weird behavior on my end, including putting endpoints into 'not-ready' state due to compilation errors when trying to regenerate the endpoints. I can reproduce this fairly easily by changing the locking logic. I checked the generated header files / C code when I saw that the endpoints couldn't regenerate and I couldn't figure out why the compiler was complaining.

ianvernon replied:
Resolved this offline with @tgraf.

tgraf commented on daemon/daemon.go:
```diff
@@ -1086,23 +1084,56 @@ func (h *patchConfig) Handle(params PatchConfigParams) middleware.Responder {
+				d.GetPolicyRepository().Mutex.RLock()
+				_, policyEnforcementChanged = d.EnablePolicyEnforcement()
+				d.GetPolicyRepository().Mutex.RUnlock()
+				d.TriggerPolicyUpdates([]policy.NumericIdentity{})
```
You can move the `TriggerPolicyUpdates()` to the `if changes > 0` branch

ianvernon replied:
That means that upon any configuration update to the daemon, we might have to regenerate each endpoint as opposed to only if PolicyEnforcement is changed - is this desirable? I figured we'd keep it here so we don't have to regenerate if needed, as regeneration is costly.

aanm commented on daemon/policy.go:
```diff
@@ -77,46 +76,65 @@ func (d *Daemon) TriggerPolicyUpdates(added []policy.NumericIdentity) *sync.Wait
 
 // EnablePolicyEnforcement returns whether policy enforcement needs to be
-// enabled for the daemon.
+// enabled at the daemon-level.
```
Can you update this comment? it returns 2 booleans, not one.

ianvernon replied:
This is in one of the earlier commits in the branch - I have reverted this functionality in a later commit. I'm going to re-order the commits in the branch, and will re-push once that's done

aanm commented on daemon/policy.go:
```diff
@@ -77,46 +76,65 @@ func (d *Daemon) TriggerPolicyUpdates(added []policy.NumericIdentity) *sync.Wait
 		}
 	}
-	return false
+	config[endpoint.OptionPolicy] = "disabled"
```
A comment to known when would this line be reached would help.

ianvernon replied:
Comment has been added.

aanm commented on daemon/policy.go:
```diff
@@ -77,46 +76,65 @@ func (d *Daemon) TriggerPolicyUpdates(added []policy.NumericIdentity) *sync.Wait
 	}
-
-	d.GetPolicyRepository().Mutex.RLock()
-	d.EnablePolicyEnforcement()
```
Why was this removed?

ianvernon replied:
It's out of place here given how this patch changes how we handle configuration updates. his is handled as part of the handler for configuration changes, which calls `d.EnablePolicyEnforcement` before it calls `TriggerPolicyUpdates`. Since `TriggerPolicyUpdates` is triggered when the daemon's configuration is changed for `PolicyEnforcement`. `EnablePolicyEnforcement()` doesn't actually change any configuration in the daemon any more in this patch, so running it has no effect here anyway.

aanm commented on pkg/endpoint/owner.go:
```diff
@@ -27,15 +27,12 @@ type Owner interface {
-	// PolicyEnabled returns whether policy enforcement is enabled
-	PolicyEnabled() bool
-
 	// EnablePolicyEnforcement returns whether owner should enable policy enforcement.
```
Update comment accordingly with the number of boolean returned

ianvernon replied:
This is no longer relevant with the latest commits for this branch.

aanm commented on tests/k8s/tests/00-gsg-test.sh:
```diff
@@ -25,10 +25,10 @@ K8SDIR="${dir}/../../../examples/kubernetes"
-	kubectl delete -f "${MINIKUBE}/l3_l4_policy_deprecated.yaml" 2> /dev/null || true
-	kubectl delete -f "${MINIKUBE}/l3_l4_policy.yaml" 2> /dev/null || true
-	kubectl delete -f "${GSGDIR}/demo.yaml" 2> /dev/null || true
+  kubectl delete -f "${MINIKUBE}/l3_l4_l7_policy.yaml" 2> /dev/null || true
```
This could be made on a different PR, it's not easy to review 30 commits on a single PR

aanm commented on tests/helpers.bash:
```diff
@@ -235,6 +235,17 @@ function wait_for_running_pod {
 
+function wait_for_no_pods {
+  set +x
+  namespace=${2:-default}
```
no_pods=$1

aanm commented on tests/helpers.bash:
```diff
@@ -235,6 +235,17 @@ function wait_for_running_pod {
+  namespace=${2:-default}
+  echo "Waiting for no pods to be Running in namespace ${namespace}"
+  kubectl get pods -n ${namespace} -o wide
+  while [[ "$(kubectl get pods -n ${namespace} -o wide 2>&1 | grep -c 'No resources found')" -ne "1" ]] ; do
```
`-ne "${no_pods}"`

ianvernon replied:
line 240 should be `namespace=${1:-default}`, but I don't see the need for ${no_pods}. On line 243, I am checking that the number of lines containing "No resources found" is 1, which is what is shown when there are no pods running in a namespace.

aanm commented on tests/14-policy-enforcement-docker.sh:
```diff
@@ -87,7 +87,7 @@ function import_sample_policy {
      "endpointSelector": {"matchLabels":{"k8s:id":"app3"}},
      "egress": [{
-         "toCIDR": [ { "ip": "9.9.9.9/32" } ]
+         "toCIDR": [ "9.9.9.9/32" ]
```
This can be squashed with ab2dd4ead4fb30ab4baeb5acd32dd13b19d80c8e (2 commits before)

ianvernon replied:
I've squashed the test changes into one commit.

Commit:
api/v1: add CIDRPolicy, PolicyEnforcement

* Add CIDRPolicy ingress and egress rules to the EndpointPolicy API object.
* Add PolicyEnforcement to Configuration object, and send Configuration object
instead of ConfigurationMap when applying Configuration for the daemon.
* Make endpoint policy enforcement field a required field in the Endpoint API
model.
* Update cilium config command, cilium API client library to account for new
Configuration API model.

Signed-off by: Ian Vernon <ian@covalent.io>

Commit:
daemon: fix grammatical error in k8s_watcher.go log message

Signed-off by: Ian Vernon <ian@covalent.io>

Commit:
pkg/policy: GetModel for L3Policy

Generate model for L3 Policy (L3 Ingress / Egress rules)

Signed-off by: Ian Vernon <ian@covalent.io>

Commit:
cilium/cmd: add check for reserved identities in requireEndpointID

cilium bpf policy list would fail when the command was provided with 'host' or 'world' as arguments. This checks whether the provided ID is one of these reserved identities.

Signed-off by: Ian Vernon <ian@covalent.io>

Commit:
pkg/option: fix log message in error

Signed-off by: Ian Vernon <ian@covalent.io>

Commit:
PolicyEnforcement fixes

* Add a lock, EnablePolicyMU to be used whenever config.EnablePolicy
is accessed or updated.
* Add a lock, ConfigPatchMutex to be used whenever PATCH requests are
made to /config via API.
* Trigger policy updates when PolicyEnforcement configuration is updated.
* Remove Policy enforcement as a boolean option for the daemon.
* Update model generated for EndpointPolicy to contain L3Policy.

Signed-off by: Ian Vernon <ian@covalent.io>

Commit:
Add PolicyEnforcement tests

* tests/k8s/tests/00-gsg-test.sh: change spacing in cleanup function from a tab
to two spaces.
* tests/run-tests: add more whitespace between each test run.
* tests/helpers.bash: add k8s_delete_policy, wait_no_pods function.
* tests/helpers.bash: add two minute timeout to cilium policy wait commands
* tests/helpers.bash: ensure that checks for endpoints being ready does not
pass when endpoints are in 'not-ready' state.
* Update tests that configure PolicyEnforcement to reflect the changes in the
command from Policy={Enabled,Disabled} to
PolicyEnforcement={always,never,default}.
* tests/14-policy-enforcement-docker: tests each possible transition in state or
policy enforcement, with and without policies imported and applying to
endpoints. It ensures that the PolicyEnforcement flag for each endpoint is
correctly updated when the daemon's PolicyEnforcement configuration flag is
changed and when a policy is imported or deleted.
* tests/k8s: add test for PolicyEnforcement like
tests/14-policy-enforcement-docker, except running in a K8s environment.
* tests/k8s/helpers.bash: get logs of previous pods in case pods have crashed.

Signed-off by: Ian Vernon <ian@covalent.io>

ianvernon commented on daemon/daemon.go:
```diff
@@ -322,10 +327,24 @@ func (d *Daemon) writeNetdevHeader(dir string) error {
+// Must be called with d.conf.EnablePolicyMU locked.
+func (d *Daemon) fmtPolicyEnforcement() string {
+	d.GetPolicyRepository().Mutex.RLock()
+	enforcement := d.EnablePolicyEnforcement()
```
Note: this always returns false in the case of when Cilium is running in tandem with K8s in 'default' mode, meaning that we always allow egress traffic to the host. Discussed offline with @tgraf , and we both agree that this isn't trivial to deal with at this time.

tgraf commented on api/v1/openapi.yaml:
```diff
@@ -702,6 +703,8 @@ definitions:
       policy-revision:
         description: The policy revision this endpoint is running on
         type: integer
+      cidr-policy:
```
Can we move this into `EndpointPolicy` instead?

tgraf commented on daemon/policy.go:
```diff
@@ -77,45 +76,59 @@ func (d *Daemon) TriggerPolicyUpdates(added []policy.NumericIdentity) *sync.Wait
+
+	config.EnablePolicyMU.RLock()
+	daemonPolicyEnable := d.EnablePolicyEnforcement()
+	defer config.EnablePolicyMU.RUnlock()
```
Can you remove the defer one line up?

tgraf reviewed (changes requested):
Looks good otherwise
